web/blueprints/finance/__init__.py

```diff
@@ -412,81 +411,106 @@ def bank_account_activities_edit(activity_id):
+    form = _create_combined_form(field_list_user, field_list_team)
 
+    return render_template('finance/bank_accounts_match.html', form=form,
+                           activities_user=matched_activities_user,
+                           activities_team=matched_activities_team)
+
+def _create_field_list_and_matched_activities_dict(matching, prefix):
     matched_activities = {}
-    for activity, user in matching.items():
-        matched_activities[str(activity.id)] = {
+    field_list = []
+    for activity, entity in matching.items():
+        matched_activities[f"{prefix}-{str(activity.id)}"] = {
             'purpose': activity.reference,
             'name': activity.other_name,
-            'user': user,
+            prefix: entity,
             'amount': activity.amount
         }
-        FieldList.append((str(activity.id), BooleanField(str(activity.id),
-                                                         default=True)))
+        field_list.append((str(activity.id), BooleanField(str(activity.id), default=True)))
 
-    class F(forms.ActivityMatchForm):
-        pass
-
-    for (name, field) in FieldList:
-        setattr(F, name, field)
-    form = F()
-
-    return render_template('finance/bank_accounts_match.html', form=form,
-                           activities=matched_activities)
+    return field_list, matched_activities
 
 
 @bp.route('/bank-account-activities/match/do/', methods=['GET', 'POST'])
 @access.require('finance_change')
 def bank_account_activities_do_match():
     # Generate form again
-    matching = match_activities()
+    matching_user, matching_team = match_activities()
+
+    field_list_user = _create_field_list(matching_user)
+    field_list_team = _create_field_list(matching_team)
+    form = _create_combined_form(field_list_user, field_list_team)
+
+    matched_user = []
+    matched_team = []
+    if form.user.validate_on_submit() or form.team.validate_on_submit():
+        # parse data
+        matched_user = _apply_checked_matches(matching_user, form.user)
+        matched_team = _apply_checked_matches(matching_team, form.team)
+        end_payment_in_default_memberships(current_user)
 
-    matched = []
-    FieldList = []
-    for activity, user in matching.items():
-        FieldList.append(
-            (str(activity.id), BooleanField('{} ({}€) -> {} ({}, {})'.format(
-                activity.reference, activity.amount, user.name, user.id,
-                user.login
+        session.flush()
+        session.commit()
+
+    return render_template('finance/bank_accounts_matched.html', matched_user=matched_user,
+                           matched_team=matched_team)
+
+def _create_field_list(matching):
+    field_list = []
+    for activity, entity in matching.items():
+        field_list.append(
+            (str(activity.id), BooleanField('{} ({}€) -> {} ({})'.format(
+                activity.reference, activity.amount, entity.name, entity.id
             ))))
 
+    return field_list
```

Comment:
This would be a good use case for a generator function:
```suggestion
def iter_matching_fields(matching: dict):
    for activity, entity in matching.items():
        yield str(activity.id), \
            BooleanField('{} ({}€) -> {} ({})'.format(
                activity.reference, activity.amount, entity.name, entity.id
            ))
```

And then use it above with `field_list_user = list(iter_matching_fields(…)`

Reply:
Also, you pass all of `matching` even though you only need `matching.items()`!
A good metaphor for this principle is “you don't give someone a wallet for them to take out the money, you give someone the money they need”.

Not particularly relevant here, but a good principle to obey for building maintainable code.